DETAILS.md

```diff
@@ -0,0 +1,115 @@
+# Details on Athens functionality.
+
+This document describes the various use cases that Athens supports or will
+support. Not everything here is implemented.
+
+[@bketelsen](https://github.com/bketelsen) and I spoke at length on Slack about
+how the Athens registry should work. I'm putting our discussion down on "paper"
+and adding some of my own detail (and some opinions) so we can grow this into
+a definitive document.
+
+# Where Will the Code Come From?
+
+Here's what was discussed on slack + various other forums:
+
+## Travis or Other CI systems (Webhook)
+
+After code passes tests, the publisher can hit a webhook to tell the registry to fetch
+code for the given tag from the given repository. The webhook will then do the following:
+
+* Check if the repository is already "known" (the owner of the repository will need to register it prior) - Fail if it isn't
+* Check if the given tag already exists (we want versions to be immutable) - Fail if it does
+* Download the code from the given tag and store it in a CDN
+
+We may need to require users to be authorized to hit the webhooks.
+
+## Manual Uploads
+
+This works similarly to the webhook flow, except a user would upload their module --
+including code -- manually to the registry. When they upload, they'll need to
+specify the module name and version, and obviously include the zip file with their
+source code.
+
+The zip file will need to have the `go.mod` file in it, and the server will
+parse it out.
+
+Athens will ship with a CLI to construct and upload the zip file.
+
+# How Will Modules Be Served?
+
+There are two modes of serving modules. One involves using the `GOPROXY` environment variable,
+and the other doesn't.
+
+## Using `GOPROXY`
+
+When a user sets `GOPROXY` on their system, the `vgo` tool automatically prefixes all modules
+with that value when it makes requests. For example,
+`GOPROXY=localhost:3000 go get my/thing` will execute the download protocol against
+`localhost:3000`.
+
+With this in mind, we'll give users two `GOPROXY` related workflows.
+
+### Workflow I - Local Development
+
+We'll give users a CLI that runs a proxy server that passes through to a tree on their
+local filesystem. The workflow will be the following:
+
+```console
+$ athens run local-proxy --root /someplace --vcs git
+$ GOPROXY=localhost:1234 go get my/module
+```
+
+The `go get` command executes the download protocol on the local proxy, which simply mirrors
+to the files in `/someplace`. In the case of `my/module`, it mirrors to `/someplace/my/module`, requires the `go.mod` to be there, and uses `git` (specified by `--vcs`) to list
+tags, etc...
```

Comment:
vcs might not be necessary if we're just caching locally fetched artifacts.  let's discuss more.

Reply:
I'm starting to think that the local proxy isn't even valuable, if you can set `GOPROXY` to a `file:///` URL. need to verify